core/src/main/java/com/datastax/oss/driver/api/core/data/CqlVector.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.api.core.data;
+
+import com.datastax.oss.driver.api.core.type.codec.TypeCodec;
+import com.datastax.oss.driver.shaded.guava.common.base.Preconditions;
+import com.datastax.oss.driver.shaded.guava.common.base.Predicates;
+import com.datastax.oss.driver.shaded.guava.common.base.Splitter;
+import com.datastax.oss.driver.shaded.guava.common.collect.Iterables;
+import com.datastax.oss.driver.shaded.guava.common.collect.Streams;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Representation of a vector as defined in CQL.
+ *
+ * <p>A CQL vector is a fixed-length array of non-null numeric values. These properties don't map
+ * cleanly to an existing class in the standard JDK Collections hierarchy so we provide this value
+ * object instead. Like other value object collections returned by the driver instances of this
+ * class are not immutable; think of these value objects as a representation of a vector stored in
+ * the database as an initial step in some additional computation.
+ *
+ * <p>While we don't implement any Collection APIs we do implement Iterable. We also attempt to play
+ * nice with the Streams API in order to better facilitate integration with data pipelines. Finally,
+ * where possible we've tried to make the API of this class similar to the equivalent methods on
+ * {@link List}.
+ */
+public class CqlVector<T extends Number> implements Iterable<T> {
+
+  /**
+   * Create a new CqlVector containing the specified values.
+   *
+   * @param vals the collection of values to wrap.
+   * @return a CqlVector wrapping those values
+   */
+  public static <V extends Number> CqlVector<V> newInstance(V... vals) {
+
+    // Note that Array.asList() guarantees the return of an array which implements RandomAccess
+    return new CqlVector(Arrays.asList(vals));
+  }
+
+  /**
+   * Create a new CqlVector that "wraps" an existing ArrayList. Modifications to the passed
+   * ArrayList will also be reflected in the returned CqlVector.
```

Comment:
It would be good to have some tests around this behavior, along with similar tests that modifying the vals var-args array in the other constructor is not reflected in the CqlVector

Reply:
Agreed.  There's a general need for a CqlVectorTest to validate behaviours for the value type itself.  That's next on my list.

Reply:
This has been addressed by the newly-added CqlVectorTest